README.md
### Willkommen beim Calibre-Autoupdater ###

Calibre ist ein Programm zum verwalten von eBooks. Es kann aber noch viel mehr. Informationen findest Du auf der [Calibre Webseite](http://www.calibre-ebook.com).
Der calibre-autoupdater schaut bei jedem Aufruf ob eine neue Calibre Version vorhanden ist und aktuallisiert sie bei bedarf. Er macht aber noch mehr. Du kannst damit auch Calibre deinstallieren.
Das Installationsscript vom Autoupdater kann Calibre bei bedarf installieren.Die Installation von Calibre sowie des Updaters erfolgt in einem von angegebenen Ort.
Du kannst den Updater dann mittels cron oder beim Autostart deiner Desktopumgebung automatisch jeden Tag starten lassen.

# Zusätzlich benötigte Programme #
  1. url, wget
  1. otify-send (nicht unbedingt nötig, aber schone Sache :-) )
  1. rep and sed (sind Linux Bordmittel und sollten dabei sein)

## Optionen für calibre-autoupdate ##

  * h, --help                   **zeigt die Hilfe an**
  * v, --version                **gibt die Programmversion aus**
  * c, --check                  **kontrolliert ob benötigte Programme vorhanden sind**
  * u, --uninstall              **deinstalliert den Calibre-Autoupdater**
  * --uninstall\_calibre      **deinstalliert das Calibre Programm**

## Optionen für calibre-autoupdate Installer (CAU\_Installer) ##
  * h, --help              **zeigt die Hilfe an**
  * V, --version           **gibt die Programmversion aus**
  * i, --install           **installiert den Calibre-Autoupdater (CAU)**
  * --install\_calibre   **installiert Calibre in der aktuellsten Version**


ChangeLog.md
# Übersicht über die Versionshistorie #

### Versionsverlauf ###
**v0.1**
  * Body des Skriptes erstellt
  * Abfrage neuste Version vorhanden
  * Abfrage welche Version installiert
  * Funktion Versionsvergleich
  * Download und installation der neusten Version
  * Desktopmeldungen hinzugefügt

**v0.2**
  * Verbesserungen bei den Desktopmeldungen

**v0.3**
  * Funktion Check Verfügbarkeit der Downloadseiten hinzugefügt

**v0.4**
  * Check Verfügbarkeit auf Test Netzwerk vorhanden und Auswertung erweitert
  * diverse Kommentare hinzugefügt

**v0.5**
  * Funktion Check läuft noch eine Calibre Instanz mit Counter
  * Ausgabe Texte verbessert und erweitert
  * URL Statuscodes holen ins main verlagert.

**v0.6**
  * Verbesserungen in der Terminalausgabe, Startausgabe ist eine eigene Funktion
  * URL Status Code ist nun eine eigene Funktion

**v0.7**
  * Der Updateprozes wurde in eine Prozedure geschrieben und die main Prozedure gibt nur noch Hilfe und Verion aus oder startet sofort die Update Prozedur

**v0.8**
  * Optionsschalter hinzugefügt, Kontrolle ob benötigte Programme vorhanden sind.
  * Unterschiedliche Installationsart und -dialog je nach dem ob man Schreibrechte
  * auf im Calibre Installationspfad hat oder nicht.

**v0.9**
  * einbinden eines Konfigurationsfiles, auslagern einiger Funktionen in Headerfiles
  * einbringen der Dateien in die Linux Verzeichnishierarchie

**v0.10**
  * hinzufügen eines Installationsskriptes für den calibre-autoupdater, Installsript prüft vorhanden sein von calibre und des updaters und fragt nach Installationsort des Updaters

**v0.11**
  * Das Mainscript wurde angepasst so das der Installer die ermittelten Variablen für den Installationsort in das Mainscript und die Konfigdatei schreiben kann.
  * Es wurden diverse Ausgabentexte sprachlich verbessert und Fehler beseitigt.

**v0.13**
  * Es wurden diverse Fehler beim eigentlichen Updateprozess bereinigt.
  * Desweiteren wurden größere Prozesse in kleine aufgeteilt.

**v0.14**
  * Uninstallfunktion für Calibre hinzugefügt. Hiermit kann Calibre von der Festplatte deinstalliert werden.
  * Bei einer nicht vorhandenen Calibreinstallation wird vom Installsript des Updaters gefragt, ob Calibre installiert werden soll.
  * Die Funktion Installation des Updaters wurde aufgeteilt in root\_install und noroot\_install
  * Uninstall Funktion für den calibre-autoupdater mit gleichzeitiger nachfrage und eventueller deinstallation vom Calibre Programm
  * der Code wurde überarbeitet, die Dialoge verbessert, Ergebnisse von Funktionsaufrufen in Variablen geschrieben

**v0.20**
  * Der Code wurde komplett überarbeitet und vieles neu geschrieben. Sauberer Code und weniger Abfragen.
  * Die Bedingungen und Schleifen wurden einfacher und logischer geschrieben
  * Die Prüfung ob Calibre noch gestartet ist wenn ein Update gemacht wird ist nun simpler geschrieben, und es wird nicht mehr hart beendet sondern gewartet bis der Benutzer Calibre beendet
  * Die Calibre Installation ist vom Installer in das Updaterskript gewandert, und im Zuge einer Calibreinstallation wird automatisch die Konfigurationsdatei des Updaters angepasst (übergabe Calibre-Install-Ort)
  * Es wird wieder geprüft ob Calibre installiert ist bevor der Updater gestartet wird

**TODO**
  * Auslesen der Download URL von der Download Seite kontrolle ob alle benötigten programme für das skript vorhanden sind


CAU_Installer.h
#  Calibre Auto Updater Header File
#    Copyright (C) 2013-2015  Leon Gaultier
#

func_term_output () {			# einheitliche Terminalausgabe, Terminal löschen
    ## Beginne mit der Ausgbe ##
    tty -s && clear
    echo -e "\033[44m \033[0m\033[1m $PROGNAME: Der Calibre-Autoupdate-Installer Version $VERSION\033[0m\n\033[44m \033[0m   2013-2015 by Leon Gaultier\n\e[m";
    return 0
}

func_usage () {    # Gibt Auskunft über die Aufrufmöglichkeiten des Scriptes
    func_term_output
    echo "Verwendung: $PROGNAME [OPTION...]"
    echo -e "Verwende $PROGNAME --help oder $PROGNAME -h für mehr Informationen.\n"
    return 0
}

func_help () {     # Listet alle Optionsschalter auf
    func_term_output
    echo "Verwende: $PROGNAME [OPTION...]"
    echo "$PROGNAME -- Der Calibre-Autoupdate-Installer (CAUI)"
    echo ""
    echo "Options:"
    echo "   -h, --help              zeigt die Hilfe an"
    echo "   -V, --version           gibt die Programmversion aus"
    echo "   -i, --install           installiert den Calibre-Autoupdater (CAU)"
    echo ""
    echo -e "Sende Fehler an <leongaultier@gmail.com>.\n"
    return 0
}

func_version () {  	# Versionsauskunft des Updater Installers und durch Versionsnummernübergabe an das Updaterprogramm auch Version des Updater Scriptes

    func_term_output
    echo "$PROGNAME: Der Calibre-Autoupdate-Installer (CAUI) v$VERSION"
    echo "Leon Gaultier 2013-2014. Das Programm darf verteilt werden,"
    echo -e "in der Hoffnung das es hilfreich ist, aber OHNR IRGENDEINE GARANTIE\n"
    return 0
}

func_progressbar() {		# Funktion Fortschrittsanzeige, Fake - Joke
    echo -e -n $'\033[34m------------------------------------------------------------------------- 100%\r'
    echo -e -n "\033[32m0% "
    for ((i=0; i<61; i++)); do
	echo -e -n "\033[32m="  
	case $i in
	    15)
		echo -e -n "25%" ;;
	    30)
		echo -e -n "50%" ;;
	    45)
		echo -e -n "75%" ;;
	    60)
		echo -e -n " 100%" ;;
	esac
    sleep 0.07
    done
    echo -e "\e[m"
    return 0
}


calibre-autoupdate_checks.h
#  Calibre Auto Updater Configfile
#    Copyright (C) 2013-2015  Leon Gaultier
#

func_http_status_code () {
    echo -e "\033[1;34m Ermittle Status von $CHECK_CALIBRE_DOWNLOAD_PAGE. Bitte warten.\e[m"
    stat_1=$(curl -o /dev/null --silent --head --write-out '%{http_code}' $CHECK_CALIBRE_DOWNLOAD_PAGE)
    func_progressbar
    echo -e "\033[32m Abgeschlossen\e[m"
    echo -e "\n\033[1;34m Ermittle Status Code von $CHECK_DOWNLOAD_URL. Bitte warten.\e[m"
    stat_2=$(curl -o /dev/null --silent --head --write-out '%{http_code}' $CHECK_DOWNLOAD_URL)
    func_progressbar
    echo -e "\033[32m Abgeschlossen\e[m"
    return 0
}

func_check_stat () {    # Funktion zum Check Verfügbarkeit der Downloadseiten und der Internetverbindung
    # Test for network conection
    echo -e "\n\033[1;34m Suche nach vorhandenen Netzwerk für die Verbindung zum Internet :-)\e[m"
    for INTERFACE in $(ls /sys/class/net/ | grep -v 'lo\|vir*'); do
        if [[ $(cat /sys/class/net/$INTERFACE/carrier) = 1 ]] && [[ -n $(ifconfig $INTERFACE | grep 'inet Adresse:' | cut -d: -f2 | awk '{ print $1 }') ]]; then
            ONLINE=1
            IFACE=$INTERFACE
        fi
    done
    func_progressbar
    if [ $ONLINE ]; then
        echo -e "\033[32m Es wurde ein Netzwerkinterface gefunden. Verbindung über Interface $IFACE\n\e[m"
        func_http_status_code
        if [[ $stat_1 -eq 200 && $stat_2 -eq 200 ]]; then
            return 1
        elif [[ $stat_1 -eq !200 ]]; then
            return 2
        else
            return 3
        fi
    else
        echo -e "\n\033[31m Fehler!!! Bitte schau Dir die Desktop Benachrichtigung an!\e[m"
        return 4
    fi
    return 0
}

func_check_run_calibre () {
    while [ `ps ax | grep /opt/calibre/bin/calibre | grep -v grep | awk '{printf $1}'` ]; do
	$NOTIFY "Um das Update installieren zu können, muss Calibre geschlossen werden..."
    done
    return 0
}

func_check_version_alternate() {	# Wurde von einem User im Mobilreaderforum vorgeschlagen. Ist mir aber ehrlich
					# gesagt zu unflexibel. Habe es für mich selber aber mal Dokumentiert
    calibre-debug -c "import urllib as u; from calibre.constants import numeric_version; raise SystemExit(int(numeric_version  < (tuple(map(int, u.urlopen('http://calibre-ebook.com/downloads/latest_version').read().split('.'))))))"

    UP_TO_DATE=$?
    if [ $UP_TO_DATE = 0 ]; then
	return 0
    else
	return 2
    fi
    echo -e "\n\033[31m Fehler!!! Es wurde beim Versionscheck ein unbekannter Wert zurückgegeben!\e[m"
    exit 0
}

func_check_version () {
    #Ermitteln der letzten aktuell verfügbaren Version
    LATEST_VERSION=`curl -s $CALIBRE_LAST_VERSION_PAGE`
    #Ermitteln der aktuell installierten Version.
    CURRENT_VERSION=`calibre --version | sed 's/[^0-9.]*\([0-9.]*\).*/\1/'`

    if [[ $CURRENT_VERSION == $LATEST_VERSION ]]; then
	return 0
    fi
    local IFS=.
    local i ver1=($CURRENT_VERSION) ver2=($LATEST_VERSION)
    # schreibe Nullen in leere Felder von ver1
    for ((i=${#ver1[@]}; i<${#ver2[@]}; i++)); do
	ver1[i]=0
    done
    for ((i=0; i<${#ver1[@]}; i++)); do
	if [[ -z ${ver2[i]} ]]; then
	    # schreibe Nullen in leere Felder von ver2
	    ver2[i]=0
	fi
	if ((10#${ver1[i]} > 10#${ver2[i]})); then
	    return 1
	fi
	if ((10#${ver1[i]} < 10#${ver2[i]})); then
	    return 2
	fi
    done
    return 0  
}

func_check_prog () {         # Schaut ob Calibre und benötigte Programme installiert sind
    func_term_output
    if [ ! -d $CALIBRE_INSTALL_LOCATION/calibre ]; then
	$NOTIFY "Check auf Calibre-Updates nicht möglich! Calibre ist gar nicht installiert. Bitte verwende calibre-autoupdate --install-calibre um Calibre zu installieren"
	exit 0
    fi
    if [ ! -f $CURL ]; then
	$NOTIFY "Du hast Curl nich installiert, es wird aber für die Updatefunktion benötigt"
	exit 0
    fi
    if [ ! -f $AWK ]; then
	$NOTIFY "Du hast AWK nich installiert, es wird aber für die Updatefunktion benötigt"
	exit 0
    fi
    if [ ! -f /usr/bin/kdialog ]; then
	echo -e "\e[0;33mDu hast kdialog nich installiert, das Programm ist optional und wird\e[m"
	echo -e "für die Upatefunktion nicht unbedingt benötigt\n\e[m"
	sleep 10
    fi
    echo -e "\033[32mAlle benötigten Programme sind installiert. Starte $PROGNAME ohne Optionen und"
    echo -e "es wird nach einer aktuellen Calibre Version gesucht\n\e[m";
    return 0
}


calibre-autoupdate.h
#  Calibre Auto Updater Header File
#    Copyright (C) 2013-2015  Leon Gaultier
#

func_term_output () {
    ## Beginne mit der Ausgbe ##
    tty -s && clear
    echo -e "\033[44m \033[0m\033[1m $PROGNAME Version $VERSION\033[0m\n\033[44m \033[0m   2013-2015 by Leon Gaultier\n\e[m";
    return 0
}

func_usage () {
    func_term_output
    echo "Verwendung: $PROGNAME [OPTION...]"
    echo "Verwende '$PROGNAME --help' oder '$PROGNAME -h' für mehr Informationen."
    return 0
}

func_help () {
    func_term_output
    echo "Verwende: $PROGNAME [OPTION...]"
    echo "$PROGNAME -- Der Calibre! Linux Edition calibre-autoupdater (CAU)"
    echo ""
    echo "Options:"
    echo "   -h, --help                   zeigt die Hilfe an"
    echo "   -V, --version                gibt die Programmversion aus"
    echo "   -c, --check                  kontrolliert ob benötigte Programme vorhanden sind"
    echo "   -u, --uninstall              deinstalliert den Calibre-Autoupdater"
    echo "       --install-calibre        installiert Calibre in der aktuellsten Version"
    echo "       --uninstall-calibre      deinstalliert das Calibre Programm"
    echo ""
    echo -e "Sende Fehler an <leongaultier@gmail.com>.\n";
    return 0
}

func_version () {
    func_term_output
    echo "$PROGNAME: Der Calibre! autoupdater v$VERSION"
    echo "Leon Gaultier 2013-2014. Das Programm darf verteilt werden,"
    echo -e "in der Hoffnung das es hilfreich ist, aber OHNR IRGENDEINE GARANTIE\n"
    return 0
}

func_progressbar() {
    echo -e -n $'\033[34m------------------------------------------------------------------------- 100%\r'
    echo -e -n "\033[32m0% "
    for ((i=0; i<61; i++)); do
	echo -e -n "\033[32m="  
	case $i in
	    15)
	      echo -e -n "25%" ;;
	    30)
	      echo -e -n "50%" ;;
	    45)
	      echo -e -n "75%" ;;
	    60)
	      echo -e -n " 100%" ;;
	esac
    sleep 0.03
    done
    echo -e "\e[m"
    return 0
}

func_uninstall() {
    func_term_output
    while ((!gueltig)); do    # beginn der Ja/Nein Abfrage
	echo -e "\033[1;34m"
	read -sn1 -t 30 -p "Möchtest Du den Calibre-Autoupdater wirklich DeInstallieren?  [j/n] " answer
	echo -e "\e[m"
	case "$answer" in
	    [JjYy])    result=1; gueltig=1 ;;
	    [Nn])      result=0; gueltig=1 ;;
	    "")        result=0; gueltig=1 ;;
	    *)         gueltig=0 ;;
	esac
    done
    if (( ! result )); then
	echo -e "\033[1;31mDie DeInstallation des Calibre-Autoupdaters wurde durch Dich abgebrochen.\e[m\n"
	exit 0
    fi
    echo -e "\033[1;31mDer Calibre-Autoupdater wird nun deinstalliert.\e[m"
    if [ -w "$CALIBRE_INSTALL_LOCATION" ]; then
	rm -rf $INC"calibre-autoupdate.h" $INC"calibre-autoupdate_checks.h"; >&2
	rm -rf $ETC"calibre-autoupdate.conf"; >&2
	rm -rf $BIN"calibre-autoupdate"; >&2
    else	     	     
	sudo rm -rf $INC"calibre-autoupdate.h" $INC"calibre-autoupdate_checks.h"; >&2
	sudo rm -rf $ETC"calibre-autoupdate.conf"; >&2
	sudo rm -rf $BIN"calibre-autoupdate"; >&2
    fi   
    func_progressbar
    echo -e "\n\033[1;32mDer Calibre-Autoupdater wurde deinstalliert. Schade das Du ihn nicht weiter verwendest\n\e[m"
    return 0
}

func_uninstall_calibre() {
    func_term_output
    if [ ! -d $CALIBRE_INSTALL_LOCATION/calibre ] ; then
	echo -e "\033[1;31mEs konnte keine Calibre Installation gefunden werden. Die DeInstallation wurde abgebrochen\e[m\n"
	exit 0
    fi
    while ((!gueltig)); do    # beginn der Ja/Nein Abfrage
	echo -e "\033[1;34mMöchtest Du Calibre wirklich DeInstallieren? Calibre kann dann nicht mehr verwendet werden!"
	read -sn1 -t 30 -p "Deine Bücher und die Calibre Bibliothek bleiben erhalten.  [j/n] " answer
	echo -e "\e[m"
	case "$answer" in
	    [JjYy])    result=1; gueltig=1 ;;
	    [Nn])      result=0; gueltig=1 ;;
	    "")        result=0; gueltig=1 ;;
	    *)         gueltig=0 ;;
	esac
    done
    if (( ! result )); then
	echo -e "\n\033[1;31mDie DeInstallation von Calibre wurde durch Dich abgebrochen.\e[m\n"
	exit 0
    fi
    echo -e "\033[1;34mCalibre wird nun deinstalliert.\e[m"    
    if [ -w "$CALIBRE_INSTALL_LOCATION" ]; then
	calibre-uninstall
    else	     	     
	sudo calibre-uninstall
    fi
    echo -e "\n\033[1;32mCalibre würde DeInstalliert! Schade es ist ein tolles Programm zur eBookverwaltung :-)\n\e[m"
    return 0
}

func_install_calibre() {
# Installationsort von Calibre
    func_term_output
    read -p "Wohin soll Calibre installiert werden? [default /opt] " CALIBRE_INSTALL_LOCATION
    if [ -z $CALIBRE_INSTALL_LOCATION ]; then
	CALIBRE_INSTALL_LOCATION=/opt
    fi
    while [ ! -d $CALIBRE_INSTALL_LOCATION ]; do
	echo -e "\n\033[1;31mDas Verzeichnis $CALIBRE_INSTALL_LOCATION existiert nicht. Bitte gib ein existierendes Verzeichnis ein!\e[m"
	read -p "Wohin soll Calibre installiert werden? [default /opt] " CALIBRE_INSTALL_LOCATION
    done
    if [ -w "$CALIBRE_INSTALL_LOCATION" ]; then
	echo -e "\n\033[1;32mCalibre wird nun installiert...\n\e[m"
	wget -nv -O- $DOWNLOAD_URL | python -c "import sys; main=lambda x:sys.stderr.write('Download failed\n'); exec(sys.stdin.read()); main('$CALIBRE_INSTALL_LOCATION')"
	if [ ! -w "$ECT" ]; then
	    echo -e "\033[1;34mDu hast kein Schreibrecht auf $ETC. Die Updater Konfigurationsdatei wird mit SUDO angepasst. Bitte gib hierzu Dein Userpasswort im Terminal ein...\n\e[m"
	    sudo sed -i "s/CALIBRE_INSTALL_LOCATION=.*/CALIBRE_INSTALL_LOCATION=$#{CALIBRE_INSTALL_LOCATION/#/\//\/}/" $ETC"calibre-autoupdate.conf"
	else
	    sed -i "s/CALIBRE_INSTALL_LOCATION=.*/CALIBRE_INSTALL_LOCATION=$#{CALIBRE_INSTALL_LOCATION/#/\//\/}/" $ETC"calibre-autoupdate.conf"
	fi
    else	     	     
	echo -e "\033[1;34mDu hast kein Schreibrecht auf $CALIBRE_INSTALL_LOCATION."
	echo -e "Calibre wird mit SUDO installiert. Bitte gib hierzu Dein Userpasswort ein...\n\e[m"
	sudo -v && wget -nv -O- $DOWNLOAD_URL | sudo python -c "import sys; main=lambda x:sys.stderr.write('Download failed\n'); exec(sys.stdin.read()); main('$CALIBRE_INSTALL_LOCATION')"
	sudo sed -i "s/CALIBRE_INSTALL_LOCATION=.*/CALIBRE_INSTALL_LOCATION=$#{CALIBRE_INSTALL_LOCATION/#/\//\/}/" $ETC"calibre-autoupdate.conf"
    fi
    echo -e "\n\n\033[1;32mHerzlichen Glückwunsch. Calibre wurde unter $CALIBRE_INSTALL_LOCATION installiert"
    echo -e "und kann nun mit \"calibre\" verwendet werden.\n\033[0m"
    return 0
}
